httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/HttpAuthenticator.java

```diff
@@ -186,52 +202,109 @@ public boolean updateAuthState(
             if (LOG.isDebugEnabled()) {
                 LOG.debug("{} Response contains no valid authentication challenges", exchangeId);
             }
-            authExchange.reset();
-            return false;
+            if (!isChallengeExpected) {
+                authExchange.reset();
+                return false;
+            }
         }
 
         switch (authExchange.getState()) {
             case FAILURE:
                 return false;
             case SUCCESS:
-                authExchange.reset();
-                break;
+                if (!isChallengeExpected) {
+                    authExchange.reset();
+                    break;
+                }
+                // otherwise fall through
             case CHALLENGED:
+                // fall through
             case HANDSHAKE:
                 Asserts.notNull(authExchange.getAuthScheme(), "AuthScheme");
+                // fall through
             case UNCHALLENGED:
                 final AuthScheme authScheme = authExchange.getAuthScheme();
+                // AuthScheme is only set if we have already sent an auth response, either
+                // because we have received a challenge for it, or preemptively.
                 if (authScheme != null) {
                     final String schemeName = authScheme.getName();
                     final AuthChallenge challenge = challengeMap.get(schemeName.toLowerCase(Locale.ROOT));
-                    if (challenge != null) {
+                    if (challenge != null || isChallengeExpected) {
                         if (LOG.isDebugEnabled()) {
-                            LOG.debug("{} Authorization challenge processed", exchangeId);
+                            LOG.debug("{} Processing authorization challenge {}", exchangeId, challenge);
                         }
                         try {
-                            authScheme.processChallenge(challenge, context);
-                        } catch (final MalformedChallengeException ex) {
+                            if (authScheme instanceof AuthScheme2) {
+                                ((AuthScheme2)authScheme).processChallenge(host, challenge, context, challenged);
+                            } else {
+                                authScheme.processChallenge(challenge, context);
+                            }
+                        } catch (final AuthenticationException | MalformedChallengeException ex) {
                             if (LOG.isWarnEnabled()) {
-                                LOG.warn("{} {}", exchangeId, ex.getMessage());
+                                LOG.warn("Exception processing Challange {}", exchangeId, ex);
                             }
                             authExchange.reset();
                             authExchange.setState(AuthExchange.State.FAILURE);
-                            return false;
+                            if (isChallengeExpected) {
+                                throw ex;
```

Comment:
Mutual auth case, throw exception, because the server has sent a 200 or equvalent code, and we will simply return the response to the client otherwise, even if the AuthExchange.STATE is Failure.